mava/systems/sac/anakin/ff_hasac.py

```diff
@@ -0,0 +1,728 @@
+        obs, env_state, buffer_state, params, opt_states, t, key = carry
+        key, act_key, learn_key = jax.random.split(key, 3)
+        # Act
+        act_state = (params.actor, obs, env_state, buffer_state, act_key)
+        (_, next_obs, env_state, buffer_state, _), metrics = scanned_act(act_state)
+
+        # Sample and learn
+        learn_state = (buffer_state, params, opt_states, t, learn_key)
+        (buffer_state, params, opt_states, _, _), losses = scanned_train(learn_state)
+
+        t += cfg.arch.num_envs * cfg.system.rollout_length
+        return (
+            LearnerState(next_obs, env_state, buffer_state, params, opt_states, t, key),
+            (metrics, losses),
+        )
+
+    # pmap and scan over explore and update_step
+    # Make sure to not do num_envs explore steps (could fill up the buffer too much).
+    explore_steps = cfg.system.explore_steps // cfg.arch.num_envs
+    pmaped_explore = jax.pmap(
+        jax.vmap(
+            lambda state: lax.scan(explore, state, None, length=explore_steps),
+            axis_name="batch",
+        ),
+        axis_name="device",
+        donate_argnums=0,
+    )
+    pmaped_update_step = jax.pmap(
+        jax.vmap(
+            lambda state: lax.scan(update_step, state, None, length=cfg.system.scan_steps),
+            axis_name="batch",
+        ),
+        axis_name="device",
+        donate_argnums=0,
+    )
+
+    return pmaped_explore, pmaped_update_step
+
+
+def run_experiment(cfg: DictConfig) -> float:
+    # Add runtime variables to config
+    cfg.arch.n_devices = len(jax.devices())
+    cfg = check_total_timesteps(cfg)
+
+    # Number of env steps before evaluating/logging.
+    steps_per_rollout = int(cfg.system.total_timesteps // cfg.arch.num_evaluation)
+    # Multiplier for a single env/learn step in an anakin system
+    anakin_steps = cfg.arch.n_devices * cfg.system.update_batch_size
+    # Number of env steps in one anakin style update.
+    anakin_act_steps = anakin_steps * cfg.arch.num_envs * cfg.system.rollout_length
+    # Number of steps to do in the scanned update method (how many anakin steps).
+    cfg.system.scan_steps = int(steps_per_rollout / anakin_act_steps)
+
+    pprint(OmegaConf.to_container(cfg, resolve=True))
+
+    # Initialize system and make learning functions.
+    (env, eval_env), networks, optims, rb, learner_state, target_entropy, logger, key = init(cfg)
+    explore, update = make_update_fns(cfg, env, networks, optims, rb, target_entropy)
+
+    actor, _ = networks
+    key, eval_key = jax.random.split(key)
+
+    def eval_act_fn(
+        params: FrozenDict, timestep: TimeStep, key: chex.PRNGKey, actor_state: ActorState
+    ) -> Tuple[Action, Dict]:
+        keys = jax.random.split(key, eval_env.num_agents)
+        action, _ = get_actions(
+            params, actor, keys, eval_env.num_agents, eval_env.action_dim, timestep.observation
+        )
+        return action, {}
```

Comment:
why don't we do similar to ppo systems (not sure if it is the same for other sac systems), we import make_act_sac_fn from the evaluator and here we give it get actions as input

Reply:
Would be nice but it's not possible here because `get_actions` returns the actions and that function expects distributions. It would be possible if we had vmappable distributions, but I think that is for another PR.

Reply:
we decided to not do this changes, since the eval_act_fn only used for HASAC so keeping it on the system side is fine